src/MSAL.PCL/Internal/Instance/Authority.cs

```diff
@@ -0,0 +1,220 @@
+
+                string openIdConfigurationEndpoint = await this.Validate(host, tenant, callState);
+
+                //discover endpoints via openid-configuration
+                TenantDiscoveryResponse edr = await this.DiscoverEndpoints(openIdConfigurationEndpoint, callState);
+
+                if (string.IsNullOrEmpty(edr.AuthorizationEndpoint))
+                {
+                    throw new MsalServiceException(MsalError.TenantDiscoveryFailed,
+                        string.Format(CultureInfo.InvariantCulture, "Token endpoint was not found at {0}",
+                            openIdConfigurationEndpoint));
+                }
+
+                if (string.IsNullOrEmpty(edr.TokenEndpoint))
+                {
+                    throw new MsalServiceException(MsalError.TenantDiscoveryFailed,
+                        string.Format(CultureInfo.InvariantCulture, "Authorize endpoint was not found at {0}",
+                            openIdConfigurationEndpoint));
+                }
+
+                if (string.IsNullOrEmpty(edr.Issuer))
+                {
+                    throw new MsalServiceException(MsalError.TenantDiscoveryFailed,
+                        string.Format(CultureInfo.InvariantCulture, "Issuer was not found at {0}",
+                            openIdConfigurationEndpoint));
+                }
+
+                this.AuthorizationEndpoint = edr.AuthorizationEndpoint.Replace("{tenant}", tenant);
+                this.TokenEndpoint = edr.TokenEndpoint.Replace("{tenant}", tenant);
+                this.SelfSignedJwtAudience = edr.Issuer.Replace("{tenant}", tenant);
+                
+                this._updatedFromTemplate = true;
+
+                // add to the list of validated authorities so that we don't do openid configuration
+                _validatedAuthorities[this.CanonicalAuthority] = this;
+            }
+        }
+
+        protected string GetDefaultOpenIdConfigurationEndpoint()
+        {
+            var authorityUri = new Uri(this.CanonicalAuthority);
+            string host = authorityUri.Authority;
+            string path = authorityUri.AbsolutePath.Substring(1);
+            string tenant = path.Substring(0, path.IndexOf("/", StringComparison.Ordinal));
+            return string.Format(CultureInfo.InvariantCulture,
+                       "https://{0}/{1}/.well-known/openid-configuration", host, tenant);
+        }
+
+        private async Task<TenantDiscoveryResponse> DiscoverEndpoints(string openIdConfigurationEndpoint, CallState callState)
+        {
+            OAuth2Client client = new OAuth2Client();
+            return
+                await
+                    client.ExecuteRequest<TenantDiscoveryResponse>(new Uri(openIdConfigurationEndpoint),
+                        HttpMethod.Get, callState).ConfigureAwait(false);
+        }
+
+        protected abstract Task<string> Validate(string host, string tenant, CallState callState);
```

Comment:
done

---
In reply to: [77232308](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/116#discussion_r77232308) [](ancestors = 77232308)